Uppercase certain identifiers for snowflake

robdaemon opened the pull request:
Snowflake stores warehouse, database and schema names in uppper case.
When creating the connection details, uppercase them.

Resolves #9511

[ci snowflake]

Commit:
Uppercase certain identifiers for snowflake

Snowflake stores warehouse, database and schema names in uppper case.
When creating the connection details, uppercase them.

Resolves #9511

[ci snowflake]

Commit:
Only upcase if it's set

[ci snowflake]

Commit:
Snowflake db name is case sensitive

Return an error message that makes more sense if the DB cannot be found.

Commit:
update test for can-connect

returns true if it can connect, throws an exception otherwise. We throw
the exception so that the humanization of the error message can happen.

robdaemon commented:
@camsaul this is good to go. I can't do anything about the coverage diff here, since it's not including driver tests.

camsaul commented:
@robdaemon so the codecoverage is yellow only because `be-linter-cloverage` failed, and that failed because there was a test failure, not due to coverage changes. The test failure was with one of the SSH tests which failed because the port was already in use -- it's a random failure I've seen once or twice before.

camsaul commented on modules/drivers/snowflake/src/metabase/driver/snowflake.clj:
```diff
@@ -28,16 +28,26 @@
                (str account "." regionid)
-               account)]
+               account)
+        upcase-not-nil (fn [s] (when s (str/upper-case s)))]
```
I think this is going to be a problem with Turkish

robdaemon replied:
hah yes you're right, let me go change that

camsaul reviewed (changes requested):
Change `str/upper-case` to the one we have in `metabase.util` that uses `Locale/US`

Commit:
add a forced-US Locale upper-case function

Avoids potential bugs with languages that don't use Latin characters

Commit:
test new upper-case-en function

Commit:
ugh automatic addition of a namespace bit me here

Commit:
and then namespace cleanup/refactor botched it

camsaul commented on modules/drivers/snowflake/test/metabase/test/data/snowflake.clj:
```diff
@@ -45,13 +45,15 @@
    ;; `metabase.driver.snowflake`
    (when (= context :db)
-     {:db (qualified-db-name database-name)})))
+     {:db (qualified-db-name (str/lower-case database-name))})))
```
We should probably do this with `Locale/US` too. I don't know if Turkish is going to do anything weird but I'd rather not find out the hard way

camsaul commented on src/metabase/util.clj:
```diff
@@ -745,6 +745,14 @@
+(defn upper-case-en
+  "Locale-agnostic version of `clojure.string/upper-case`.
+  `clojure.string/upper-case` uses the default locale in conversions, turning
+  `ID` into `ıd`, in the Turkish locale. This function always uses the
```
I think this docstring is backwards, this function is `upper-case` but the example is lowercasing stuff.

Anyways I think we should add a `lower-case-en` version too, just in case normal `str/lower-case` with Turkish does `ID` into `ıd

camsaul commented on test/metabase/util_test.clj:
```diff
@@ -235,6 +235,16 @@
         (Locale/setDefault original-locale)))))
 
+(deftest upper-case-en-test
+  (let [original-locale (Locale/getDefault)]
+    (try
+      (Locale/setDefault (Locale/forLanguageTag "tr"))
+      ;; `(str/upper-case "id")` returns "İD" in the Turkish locale
+      (is (= "ID"
+             (u/upper-case-en "id")))
+      (finally
+        (Locale/setDefault original-locale)))))
```
we actually have a test util macro that does this already.

```clj
(mt/with-locale "tr"
  (str/lower-case "ID"))
;; -> "ıd"
```

camsaul reviewed (dismissed):
I went and checked in the REPL and it looks like `str/lower-case` does weird things too for Turkish. So we should add a `lower-case-en` function and use that instead of `str/lower-case` as well so we don't end up with more bugs when people run Metabase in Turkish

robdaemon commented:
Addressed the feedback.

We already had a `u/lower-case-en` function - I'm just adding the `u/upper-case-en` version.

I made an attempt at finding all of our `str/lower-case` and `str/upper-case` uses, but I ended up botching A LOT of stuff in the process. We should file a separate issue to do that? I do see places where things can go very wrong in the Turkish locale.

Commit:
code review feedback

flamber commented:
@robdaemon Could #11424 be expanded to include upper too?

sbelak reviewed (approved):
looks good
